plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/incremental/cache.kt

```diff
@@ -39,38 +49,70 @@ class JavaClassCacheManager(val file: File) : Closeable {
-                    )
+        val aggregatingGeneratedTypes = aptCache.getAggregatingGeneratedTypes(javaCache::getTypesForFiles)
+        val impactedTypes = getAllImpactedTypes(changes, aggregatingGeneratedTypes)
+        val isolatingGeneratedTypes = aptCache.getIsolatingGeneratedTypes(javaCache::getTypesForFiles)
+
+        val sourcesToReprocess = changedSources.toMutableSet()
+        val classNamesToReprocess = mutableListOf<String>()
+
+        if (changedSources.isNotEmpty() || impactedTypes.isNotEmpty()) {
+            for (aggregatingOrigin in aptCache.getAggregatingOrigins()) {
+                if (aggregatingOrigin in impactedTypes) continue
+
+                val originSource = javaCache.getSourceForType(aggregatingOrigin)
+                if (originSource.extension == "java") {
+                    sourcesToReprocess.add(originSource)
+                } else if (originSource.extension == "class") {
+                    // This is a generated .class file that we need to reprocess.
+                    classNamesToReprocess.add(aggregatingOrigin)
                 }
+            }
+        }
 
-                SourcesToReprocess.Incremental(
-                    toReprocess.toList(),
-                    generatedDirtyTypes,
-                    aggregatedTypes.also {
-                        it.removeAll(filesToReprocess.dirtyTypes)
-                        it.removeAll(generatedDirtyTypes)
-                        it.removeAll(invalidatedIsolatingId)
-                    })
+        for (impactedType in impactedTypes) {
+            if (impactedType !in isolatingGeneratedTypes && impactedType !in aggregatingGeneratedTypes) {
+                sourcesToReprocess.add(javaCache.getSourceForType(impactedType))
+            } else if (impactedType in isolatingGeneratedTypes) {
+                // this is a generated type by isolating AP
+                val isolatingOrigin = aptCache.getOriginForGeneratedIsolatingType(impactedType, javaCache::getSourceForType)
+                if (isolatingOrigin in impactedTypes || isolatingOrigin in dirtyClasspathFqNames) {
+                    continue
+                }
+                val originSource = javaCache.getSourceForType(isolatingOrigin)
+                if (originSource.extension == "java") {
+                    sourcesToReprocess.add(originSource)
+                } else if (originSource.extension == "class") {
+                    classNamesToReprocess.add(isolatingOrigin)
+                }
             }
         }
+
+        if (sourcesToReprocess.isNotEmpty() || classNamesToReprocess.isNotEmpty()) {
+            // Invalidate state only if there are some files that will be reprocessed
+            javaCache.invalidateDataForTypes(impactedTypes)
+            aptCache.invalidateAggregating()
+            // for isolating, invalidate both own types and classpath types
+            aptCache.invalidateIsolatingForOriginTypes(impactedTypes)
```

Comment:
Maybe `aptCache.invalidateIsolatingForOriginTypes(impactedTypes + dirtyClasspathFqNames) better?

Reply:
I wanted to avoid creating additional set, but considering we iterate over all isolating mappings, it is more efficient to do so.

Reply:
ok